packages/nimble-components/src/table/index.ts

```diff
@@ -754,9 +886,19 @@ export class Table<
                 this.selectionMode
             );
         }
-        if (this.tableUpdateTracker.requiresTanStackDataReset) {
-            // Perform a shallow copy of the data to trigger tanstack to regenerate the row models and columns.
-            updatedOptions.data = [...this.table.options.data];
+        if (
+            this.tableUpdateTracker.requiresTanStackDataReset
+            && !this.tableUpdateTracker.updateRowParentIds
+        ) {
+            if (!this.parentIdFieldName) {
+                // Perform a shallow copy of the data to trigger tanstack to regenerate the row models and columns.
+                updatedOptions.data = [...this.table.options.data];
+            } else {
+                const flatList = this.convertHierarchicalDataToFlatList();
```

Comment:
So we need to reprocess the data in the situation where:
1. `parentId` is set
2. `id` is not set
3. table is given data (this results in a flat data set in Tanstack, since there was no `id` set)
4. `id` is then set

The final step results in `requiresTanStackDataReset` being true but `updateRowParentIds` being false, and we still need to reprocess the data to convert it to its hierarchical form. It seems like there should be clearer boolean logic to handle all of the needed cases, but I haven't come up with it yet.